### PR TITLE
Add rounded-rect filter shader

### PR DESCRIPTION
**Pull request by carlosbaraza**
This shader example is inspired by [`rounded_rect.shader`](https://github.com/Oncorporation/obs-shaderfilter/blob/master/data/examples/rounded_rect.shader).

### Description
Add a shader to round corners with a filter. E.g.

Screenshot 2021-04-30 at 20 54 45

#### Old Behavior
Just an example

#### New Behavior
Just an example

### Related Issues
No related issues as far as I know.

### CODE REVIEW

File: data/examples/shaders/filter/rounded-rect.effect [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+technique Draw
+{
+	pass
+	{
+		vertex_shader = DefaultVertexShader(vtx);
+		pixel_shader  = CombinedCurvature(vtx);
+	}
+}
+
+technique Combined
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:78]: This technique doesn't seem relevant to the actual effect?

File: data/examples/shaders/filter/rounded-rect.effect [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	string field_type = "input";
+	int minimum = 0;
+	int maximum = 300;
+	int step = 1;
+	int scale = 1;
+> = 5;
+
+
+//------------------------------------------------------------------------------
+// Technique: Combined Curvature
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:39]: **Optional:** Consider renaming the technique, as I'm unsure if it applies here.

File: data/examples/shaders/filter/rounded-rect.effect [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+//------------------------------------------------------------------------------
+
+float4 CombinedCurvature(VertexInformation vtx) : TARGET {
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:42]: **Optional:** Consider renaming the pixel shader to match the actual functionality.

File: data/examples/shaders/filter/rounded-rect.effect [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	float4 rgb = InputA.Sample(LinearClampSampler, uv);
+
+    float2 pixel_position = float2(mirrored_tex_coord.x * ViewSize.x, mirrored_tex_coord.y * ViewSize.y);
+    float pixel_distance_from_center = distance(pixel_position, float2(corner_radius, corner_radius));
+    
+    bool is_in_corner = pixel_position.x < corner_radius && pixel_position.y < corner_radius;
+    bool is_within_radius = pixel_distance_from_center <= corner_radius;
+    
+    bool is_within_edge_border = !is_in_corner && (pixel_position.x <= _2_border_thickness || pixel_position.y <= _2_border_thickness);
+    bool is_within_corner_border = is_in_corner && pixel_distance_from_center < corner_radius && pixel_distance_from_center > (corner_radius - _2_border_thickness);
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:56]: It is fine to && and || statements which are a single word, but for multiple words or even comparison, the statement should be surrounded by (). This prevents situations where one Driver compiler considers it to be `(a > b) && c` and the other `a > (b && c)` because of a driver bug. 

##### Example:
```
bool bad = d || a > b && c < d;
bool ambigious = d || (a > b) && (c < d);
bool good = d || ((a > b) && (c < d));
```

Please adjust this for every instance.

File: data/examples/shaders/filter/rounded-rect.effect [hunk 5]
```diff
@@ -0,0 +1,85 @@
+#define IS_FILTER
+#include "../base.effect"
+
+//------------------------------------------------------------------------------
+// Uniforms
+//------------------------------------------------------------------------------
+
+uniform int _0_corner_radius<
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:8]: **Optional:** Prefer float, as int breaks on libobs-opengl.

File: data/examples/shaders/filter/rounded-rect.effect [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+uniform float4 _1_border<
+	string name = "Border Color";
+	string field_type = "slider";
+	float4 minimum = {0., 0., 0., 0.};
+	float4 maximum = {100., 100., 100., 100.};
+	float4 step = {0.01, 0.01, 0.01, 0.01};
+	float4 scale = {0.01, 0.01, 0.01, 0.01};
+> = {0., 0., 0., 100.};
+
+uniform int _2_border_thickness<
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:27]: **Optional:** Prefer float, as int breaks on libobs-opengl.

File: data/examples/shaders/filter/rounded-rect.effect [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+//------------------------------------------------------------------------------
+// Technique: Combined Curvature
+//------------------------------------------------------------------------------
+
+float4 CombinedCurvature(VertexInformation vtx) : TARGET {
+    float2 uv = vtx.texcoord0.xy;
+    int corner_radius = _0_corner_radius;
+
+    float2 mirrored_tex_coord = float2(0.5, 0.5) - abs(uv - float2(0.5, 0.5));
+	float4 rgb = InputA.Sample(LinearClampSampler, uv);
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:47]: Incorrect spacing.

File: data/examples/shaders/filter/rounded-rect.effect [hunk 8]
```diff
@@ -0,0 +1,85 @@
+#define IS_FILTER
```
Comment (Xaymar) [inline data/examples/shaders/filter/rounded-rect.effect:1]: License and Copyright missing.

**Review by Xaymar — CHANGES_REQUESTED** [review]
[The StreamFX project uses a 4-wide Tab for indentation, not Spaces.](https://github.com/Xaymar/obs-StreamFX/blob/master/.editorconfig#L9).

File: data/examples/shaders/filter/rounded-rect.effect (hunk 6 above)
Comment (carlosbaraza) [reply data/examples/shaders/filter/rounded-rect.effect:27]: `float` didn't seem to work because the UI completely breaks showing extremely high numbers. Even if I use `input`, it doesn't really work, like if the uniform wasn't ever defined.

File: data/examples/shaders/filter/rounded-rect.effect (hunk 8 above)
Comment (carlosbaraza) [reply data/examples/shaders/filter/rounded-rect.effect:1]: What License? Yours or mine? I don't really mind, so please feel free to add any license you want.

Comment (Xaymar) [reply data/examples/shaders/filter/rounded-rect.effect:1]: It needs your license for this file, or rather the original authors license as this is derivative work. In some areas of the world, a lack of a license header doesn't automatically put it under the same license as the rest of the code, and so far I have not yet adopted a CLA/CAA.

**Comment by carlosbaraza** [comment]
Hopefully it is a bit less sloppy this time around. I kind of did it quickly at first to get it working, and didn't pay much attention at all. Please, feel free to amend anything you might consider not up to the standards for this repo.

Regarding the `float`, I wanted to use it, but for some reason it seems to not work at all, and I couldn't find a good way to debug the problem. Please, let me know if you find a way to make it work. If I just replace the uniform for:

```
uniform float _0_corner_radius<
	string name = "Corner Radius";
	string suffix = " px";
	string field_type = "input";
	float minimum = 0.;
	float maximum = 2000.;
	float step = 10.;
	float scale = 1.;
> = 40.;
```

It will just not work, at least in Mac OSX + OBS 26.1.2 + ShaderFX 0.10.

Also noticed that the UI for `slider` doesn't work when using `int`.

**Comment by Xaymar** [comment]
> It will just not work, at least in Mac OSX + OBS 26.1.2 + ShaderFX 0.10.
> 
> Also noticed that the UI for `slider` doesn't work when using `int`.

Will have to look into this.

**Review by Xaymar — APPROVED** [review]
